File: dvm.py
#!/usr/bin/env python

from vmbase import VMBase
# This is the class that contains all the instructions for the Diddy VM

class DVM(VMBase):
    """
    This contains the methods for each of the VM instructions

        * NOP
        * Copy
        * Jump
        * Branch
        * Equal
        * Greater Than
        * Less Than
        * Add
        * Subtract
        * Output
        * Halt
    """

    def __init__(self):
        """
        The constructor of the VMBase parent class is called to set
        everything up, then the instructions dictionary is setup
        in this constructor
        """

        super(DVM, self).__init__()

        i = self.instructions

        i[0] = self.nop
        i[1] = self.copy
        i[2] = self.jump
        i[3] = self.branch
        i[4] = self.equal
        i[5] = self.greater
        i[6] = self.lesser
        i[7] = self.add
        i[8] = self.subtract
        i[9] = self.output
        i[10] = self.halt

    def nop(self):
        """
        No Operation
        Just increments the position counter
        """
        self.next()

    # Data Movement Instructions
    def copy(self):
        """
        Copy data at address A to address B
        """
        self.next()
        inAddr = self.getMem()
        inVal = self.getMem(inAddr)
        self.next()
        outAddr = self.getMem()
        self.setMem(outAddr, inVal)
        self.next()

    # Program Flow instructions
    def jump(self):
        """
        Set the next instruction pointer to the value at address A
        """
        self.next()
        inAddr = self.getMem()
        self.setInstructionPointer(inAddr)

    def branch(self):
        """
        Check the value at address A
        Set the next instruction pointer to be C if it's 0
        set it to B otherwise
        """
        self.next()
        aAddr = self.getMem()
        aVal = self.getMem(aAddr)

        self.next()
        bAddr = self.getMem()

        self.next()
        cAddr = self.getMem()

        if aVal == 0:
            self.setInstructionPointer(cAddr)
        else:
            self.setInstructionPointer(bAddr)


    # Comparison instructions
    def equal(self):
        """
        Compare two values for equality
        Write a 1 to address C if the values at address A and B
        are equal, otherwise write a 0
        """
        self.next()
        aAddr = self.getMem()
        aVal = self.getMem(aAddr)

        self.next()
        bAddr = self.getMem()
        bVal = self.getMem(bAddr)

        self.next()
        cAddr = self.getMem()

        if aVal == bVal:
            self.setMem(cAddr, 1)
        else:
            self.setMem(cAddr, 0)

        self.next()

    def greater(self):
        """
        Compare two values
        Write a 1 to address C if the value at address A
        is greater than that at address B, otherwise write a 0
        """
        self.next()
        aAddr = self.getMem()
        aVal = self.getMem(aAddr)

        self.next()
        bAddr = self.getMem()
        bVal = self.getMem(bAddr)

        self.next()
        cAddr = self.getMem()

        if aVal > bVal:
            self.setMem(cAddr, 1)
        else:
            self.setMem(cAddr, 0)

        self.next()

    def lesser(self):
        """
        Compare two values
        Write a 1 to address C if the value at address A
        is less than that at address B, otherwise write a 0
        """
        self.next()
        aAddr = self.getMem()
        aVal = self.getMem(aAddr)

        self.next()
        bAddr = self.getMem()
        bVal = self.getMem(bAddr)

        self.next()
        cAddr = self.getMem()

        if aVal < bVal:
            self.setMem(cAddr, 1)
        else:
            self.setMem(cAddr, 0)

        self.next()


    # Manipulation Instructions
    def add(self):
        """
        Add values at addresses A and B then write it to address pointed to by C
        """
        self.next()
        aAddr = self.getMem()
        aVal = self.getMem(aAddr)

        self.next()
        bAddr = self.getMem()
        bVal = self.getMem(bAddr)

        self.next()
        cAddr = self.getMem()

        value = aVal + bVal

        self.setMem(cAddr, value)

        self.next()

    def subtract(self):
        """
        Subtract value at addresses B from A then write it to address pointed to by C
        """
        self.next()
        aAddr = self.getMem()
        aVal = self.getMem(aAddr)

        self.next()
        bAddr = self.getMem()
        bVal = self.getMem(bAddr)

        self.next()
        cAddr = self.getMem()

        value = aVal - bVal

        self.setMem(cAddr, value)

        self.next()

    def output(self):
        """
        Write the value at address A to the output
        """
        self.next()
        aAddr = self.getMem()
        aVal = self.getMem(aAddr)
        self.systemOut(aVal)
        self.next()

    def halt(self):
        """
        Halt the program and set the exit status to the value at address A
        """
        self.next()
        aAddr = self.getMem()
        aVal = self.getMem(aAddr)
        self.systemExit(aVal)

if __name__ == '__main__':
    import unittest
    from tests.testdvm import TestDVM
    unittest.main()

File: cdiddy/dvm.c
#include "dvm.h"
#include "vmbase.h"

DVM setup_diddy() {

    DVM dvm = setup_vmbase();

    dvm->instructions[0] = nop;
    dvm->instructions[1] = copy;
    dvm->instructions[2] = jump;
    dvm->instructions[3] = branch;
    dvm->instructions[4] = equal;
    dvm->instructions[5] = greater;
    dvm->instructions[6] = lesser;
    dvm->instructions[7] = add;
    dvm->instructions[8] = subtract;
    dvm->instructions[9] = output;
    dvm->instructions[10] = halt;

    return dvm;
}

void nop(DVM dvm) {
    next(dvm);
}


void copy(DVM dvm) {
    next(dvm);
    int inAddr = getMem(dvm, -1);
    int inVal = getMem(dvm, inAddr);
    next(dvm);
    int outAddr = getMem(dvm, -1);
    setMem(dvm, outAddr, inVal);
    next(dvm);
}

void jump(DVM dvm) {
    next(dvm);
    int inAddr = getMem(dvm, -1);
    set_instruction_pointer(dvm, inAddr);
}

void branch(DVM dvm) {
    next(dvm);
    int aAddr = getMem(dvm, -1);
    int aVal = getMem(dvm, aAddr);

    next(dvm);
    int bAddr = getMem(dvm, -1);

    next(dvm);
    int cAddr = getMem(dvm, -1);

    if (aVal == 0) {
        set_instruction_pointer(dvm, cAddr);
    } else {
        set_instruction_pointer(dvm, bAddr);
    }
}

void equal(DVM dvm) {
    next(dvm);
    int aAddr = getMem(dvm, -1);
    int aVal = getMem(dvm, aAddr);

    next(dvm);
    int bAddr = getMem(dvm, -1);
    int bVal = getMem(dvm, bAddr);

    next(dvm);
    int cAddr = getMem(dvm, -1);

    if (aVal == bVal) {
        setMem(dvm, cAddr, 1);
    } else {
        setMem(dvm, cAddr, 0);
    }
    next(dvm);
}

void greater(DVM dvm) {
    next(dvm);
    int aAddr = getMem(dvm, -1);
    int aVal = getMem(dvm, aAddr);

    next(dvm);
    int bAddr = getMem(dvm, -1);
    int bVal = getMem(dvm, bAddr);

    next(dvm);
    int cAddr = getMem(dvm, -1);

    if (aVal > bVal) {
        setMem(dvm, cAddr, 1);
    } else {
        setMem(dvm, cAddr, 0);
    }
    next(dvm);
}

void lesser(DVM dvm) {
    next(dvm);
    int aAddr = getMem(dvm, -1);
    int aVal = getMem(dvm, aAddr);

    next(dvm);
    int bAddr = getMem(dvm, -1);
    int bVal = getMem(dvm, bAddr);

    next(dvm);
    int cAddr = getMem(dvm, -1);

    if (aVal < bVal) {
        setMem(dvm, cAddr, 1);
    } else {
        setMem(dvm, cAddr, 0);
    }
    next(dvm);
}

void add(DVM dvm) {
    next(dvm);
    int aAddr = getMem(dvm, -1);
    int aVal = getMem(dvm, aAddr);

    next(dvm);
    int bAddr = getMem(dvm, -1);
    int bVal = getMem(dvm, bAddr);

    next(dvm);
    int cAddr = getMem(dvm, -1);

    int value = aVal + bVal;

    setMem(dvm, cAddr, value);
    next(dvm);
}

void subtract(DVM dvm) {
    next(dvm);
    int aAddr = getMem(dvm, -1);
    int aVal = getMem(dvm, aAddr);

    next(dvm);
    int bAddr = getMem(dvm, -1);
    int bVal = getMem(dvm, bAddr);

    next(dvm);
    int cAddr = getMem(dvm, -1);

    int value = aVal - bVal;

    setMem(dvm, cAddr, value);
    next(dvm);
}

void output(DVM dvm) {
    next(dvm);
    int aAddr = getMem(dvm, -1);
    int aVal = getMem(dvm, aAddr);

    system_out(dvm, aVal);
}

void halt(DVM dvm) {
    next(dvm);
    int aAddr = getMem(dvm, -1);
    int aVal = getMem(dvm, aAddr);

    system_exit(dvm, aVal);
}

